helper/querybuilder.php

```diff
@@ -0,0 +1,277 @@
+<?php
+/**
+ * Tagging Plugin (helper component)
+ */
+class helper_plugin_tagging_querybuilder extends DokuWiki_Plugin {
+
+    /** @var string */
+    protected $field;
+    /** @var bool */
+    protected $logicalAnd = false;
+    /** @var array */
+    protected $tags = [];
+    /** @var array  */
+    protected $ns = [];
+    /** @var array */
+    protected $notns = [];
+    /** @var string */
+    protected $pid;
+    /** @var string */
+    protected $tagger = '';
+    /** @var string */
+    protected $limit = '';
+    /** @var string */
+    protected $where;
+    /** @var string */
+    protected $orderby;
+    /** @var string */
+    protected $groupby;
+    /** @var string */
+    protected $having = '';
+    /** @var array */
+    protected $values = [];
+
+    /**
+     * Shorthand method: calls the appropriate getter deduced from $this->field
+     *
+     * @return helper_plugin_tagging_querybuilder
+     */
+    public function getQuery()
+    {
+        if (!$this->field) {
+            throw new \RuntimeException('Failed to build a query, no field specified');
+        }
+        return ($this->field === 'pid') ? $this->getPages() : $this->getTags();
+    }
+
+    /**
+     * Processes all parts of the query for fetching tagged pages
+     *
+     * Returns the query builder object
+     *
+     * @return helper_plugin_tagging_querybuilder
+     */
+    public function getPages()
+    {
+        $this->where = $this->getWhere();
+        $this->groupby = 'pid';
+        $this->orderby = "cnt DESC, pid";
+        if ($this->tags && $this->logicalAnd) $this->having = ' HAVING cnt = ' . count($this->tags);
+
+        return $this;
```

Comment:
I think it's weird that the `getPages()` and `getTags()` methods return an instance of the querybuilder. I would expect them to actually build the query and return `[$sql, $param]`.

The way it's currently implemented, the order in which I call methods on this class matters, but is not enforced. This potentially leads to problems when I don't know the internals of this class.

Reply:
I wasn't sure if it was the right way, I just wanted to avoid returning an array with different data types. But you are right, this is worse, so the method does return the array now. It's immediately passed to a db method anyway, so its format doesn't matter much.